cmd/soroban-rpc/lib/preflight/src/fees.rs

```diff
@@ -292,118 +285,198 @@ fn storage_footprint_to_ledger_footprint(foot: &Footprint) -> Result<LedgerFootp
-    let transaction_data = SorobanTransactionData {
-        resources: soroban_resources,
-        refundable_fee: ref_fee,
-        ext: ExtensionPoint::V0,
+    finalize_transaction_data_and_min_fee(
+        &ledger_storage,
+        &transaction_resources,
+        soroban_resources,
+        &rent_changes,
+        current_ledger_seq,
+        bucket_list_size,
+    )
+}
+
+fn get_bumpable_entry_expiration_and_durability(
+    ledger_entry: &LedgerEntry,
+) -> Result<(u32, ContractDataDurability), Box<dyn error::Error>> {
+    // TODO: Are there any other acceptable entry types?
+    //       is the ContractCode entry type always considered persistent?
+    let res = match ledger_entry.data {
+        LedgerEntryData::ContractData(ref cd) => (cd.expiration_ledger_seq, cd.durability),
+        LedgerEntryData::ContractCode(ref cc) => (cc.expiration_ledger_seq, Persistent),
+        _ => {
+            let err = format!(
+                "Incorrect ledger entry type ({}) in footprint",
+                ledger_entry.data.name()
+            )
+            .into();
+            return Err(err);
+        }
     };
-    Ok((transaction_data, min_fee))
+    Ok(res)
 }
 
 pub(crate) fn compute_restore_footprint_transaction_data_and_min_fee(
     footprint: LedgerFootprint,
-    snapshot_source: &ledger_storage::LedgerStorage,
+    ledger_storage: &ledger_storage::LedgerStorage,
     bucket_list_size: u64,
+    current_ledger_seq: u32,
 ) -> Result<(SorobanTransactionData, i64), Box<dyn error::Error>> {
+    let ConfigSettingEntry::StateExpiration(state_expiration) =
+        ledger_storage.get_configuration_setting(ConfigSettingId::StateExpiration)?
+        else {
+            return Err(
+                "get_fee_configuration(): unexpected config setting entry for StateExpiration key".into(),
+            );
+        };
+    // FIXME: we are querying the ledger entries from the DB twice (once here and again in calculate_unmodified_ledger_entry_bytes)
+    let mut rent_changes: Vec<LedgerEntryRentChange> = Vec::new();
+    for key in footprint.read_write.as_vec() {
+        let unmodified_entry = ledger_storage.get(key, true)?;
+        let (expiration_ledger, durability) =
+            get_bumpable_entry_expiration_and_durability(&unmodified_entry)?;
+        if durability != Persistent {
+            let err = format!("Non-persistent key ({:?}) in footprint", key).into();
+            return Err(err);
+        }
+        // FIXME: according to https://soroban.stellar.org/docs/fundamentals-and-concepts/state-expiration :
+        //  The entry is considered expired when current_ledger > expirationLedger.
+        //  So, the code below should be `current_ledger_seq <= expiration_ledger`
+        //  However, we observe the expiration happening exactly at expiration_ledger.
+        //  Is this a bug in the ledger updates of soroban-rpc?
+        if current_ledger_seq < expiration_ledger {
```

Comment:
@sisuresh @dmkozh thoughts?

Reply:
CC @tamirms

Reply:
_we observe_ means that soroban-rpc sees the entry expire at `expiration_ledger` and not `expiration_ledger + 1`, as a result if we preflight the restoration right after the expiration and check `current_ledger_seq <= expiration_ledger` soroban-rpc considers the operation a noop and the resulting restoration transaction is not accepted.

Reply:
I still don't quite understand; the correct condition is that entry is still alive on `expiration_ledger` and expired (== inaccessible) on `expiraiton_ledger + 1`. Anything else is a bug, so IIUC soroban-rpc should be fixed to respect this logic.

On a side note, I hope that preflight sets the ledger sequence to `last_closed_ledger_seq + 1`, as that would be as close to reality as we can achieve with preflight.

Reply:
> We don't emit anything when entry expires

Aha, then we may not be expiring the entries at the right time.

Reply:
Fixed at https://github.com/stellar/soroban-tools/pull/846/commits/db65519b1b747c36701380961fbf54239cd9f7de

Reply:
Yes, the logic is correct now. But do note, that it is correct only for checking the entry state at LCL. As I've mentioned above, preflight should ideally happen at LCL + 1, as there is clearly no way to include a new transaction in LCL.